Adding memory MaxResourcePolicy to VPAs

muaazsaleem opened the pull request:
Adding memory MaxResourcePolicy to both prometheus and metrics-server
VPAs. This is so that the VPA doesn't scale prometheus and metrics-server
so much that they become unschedulable.

Note that if `promtheus_mem` is set than the vpa can't **scale up** more than
that, so this should be set to the maximum possible value in a cluster

Commit:
Moving prometheus and metrics server vpas

Moving prometheus and metrics server vpas to their manifest dirs,
so they're with other deployment specific manifests

Signed-off-by: Muhammad Muaaz Saleem <muhammad.muaaz.saleem@zalando.de>

Commit:
Adding memory MaxResourcePolicy to VPAs

Adding memory MaxResourcePolicy to both prometheus and metrics-server
VPAs. This is so that the VPA doesn't scale prometheus and metrics-server
so much that they become unschedulable.

Note that if `promtheus_mem` is set than the vpa can't *scale up* more than
that, so this should be set to the maximum possible value in a cluster

Signed-off-by: Muhammad Muaaz Saleem <muhammad.muaaz.saleem@zalando.de>

arjunrn commented on cluster/manifests/metrics-server/metrics-server-vpa.yaml:
```diff
@@ -10,4 +10,11 @@ spec:
       application: metrics-server
   updatePolicy:
     updateMode: "Auto"
+  {{if index .ConfigItems "prometheus_mem"}}
```
I think there is a default value for `prometheus_mem` so the if condition is not needed.

njuettner replied:
unfortunately not, we just only set it in the prometheus sts if the config item doesn't exist.
https://github.com/zalando-incubator/kubernetes-on-aws/blob/dev/cluster/manifests/prometheus/statefulset.yaml#L39

njuettner commented on cluster/manifests/metrics-server/metrics-server-vpa.yaml:
```diff
@@ -10,4 +10,11 @@ spec:
+    containerPolicies:
+    - containerName: metrics-server
+      maxAllowed:
+        memory: {{.ConfigItems.prometheus_mem}}
```
Wouldn't it be better to set it directly:

`memory: {{if index .ConfigItems "prometheus_mem"}}{{.ConfigItems.prometheus_mem}}{{else}}4000Mi{{end}}`

njuettner replied:
same for **prometheus-vpa.yaml**

muaazsaleem replied:
Beacause the `MaxResourcePolicy` sets the ceiling for how much the VPA can scale a container **up**, if we setup a default value e.g `4000Mi` than the VPA would never scale beyond that.i.e

```
requests <.....vpa scale up range.....> MaxResourcePolicy
```
I'm just worried that this could hurt us in large clusters ( large node sizes and high no. of nodes). 

I guess the question is, _"what is more likely?"_ The case where the VPA scales prometheus up so much that it becomes unschedulable  or that the `MaxResourcePolicy` would be set too low and the VPA won't be able to scale up `prometheus` enough? 

What do you think?

arjunrn replied:
Either way it won't be any different from the current behavior so there's no harm in doing in that way.

muaazsaleem replied:
Sounds good. I'll add a default value

aermakov-zalando replied:
I'd set the default in `config-defaults.yaml` and just use the config item in here (and set it to a much larger value, e.g. `14Gi`).

Commit:
Adding a default MaxResourcePolicy

Adding default MaxResourcePolicy(ies) to {prometheus,metrics-server}-vpa
so that the vpa doesn't scale them "infinitely".

Signed-off-by: Muhammad Muaaz Saleem <muhammad.muaaz.saleem@zalando.de>

arjunrn commented on cluster/manifests/metrics-server/metrics-server-vpa.yaml:
```diff
@@ -10,4 +10,9 @@ spec:
+    containerPolicies:
+    - containerName: metrics-server
+      maxAllowed:
+        memory: {{if index .ConfigItems "prometheus_mem"}}{{.ConfigItems.prometheus_mem}}{{else}}4000Mi{{end}}
```
Is the default for the metrics-server also 4000Mi?

njuettner replied:
it would be better to replace it with "{{.ConfigItems.metrics_service_mem}}" and then just using else whatever it's preferred.

Commit:
Fixing default MaxResourcePolicy for metrics_server

Signed-off-by: Muhammad Muaaz Saleem <muhammad.muaaz.saleem@zalando.de>

Commit:
Separating mem and max_mem config items for vpas

Separately defining max_mem config items so they can be large enough.
Also setting defaults for prometheus_mem and metrics_service_mem to
clean up the manifests

Signed-off-by: Muhammad Muaaz Saleem <muhammad.muaaz.saleem@zalando.de>

njuettner commented:
👍

arjunrn commented on cluster/config-defaults.yaml:
```diff
@@ -64,8 +64,14 @@ zmon_worker_cpu: "750m"
 
+prometheus_cpu: "1000m"
+prometheus_mem: "4Gi"
+max_prometheus_mem: "14Gi"
```
If the vpa selects this value the pods would be unschedulable in test clusters.

Commit:
Lowering max_prometheus_mem slightly

Lowering max_prometheus_mem slightly to be sure it'll fit most nodes

Signed-off-by: Muhammad Muaaz Saleem <muhammad.muaaz.saleem@zalando.de>

Commit:
Making config items more grep-able

Signed-off-by: Muhammad Muaaz Saleem <muhammad.muaaz.saleem@zalando.de>

aermakov-zalando commented:
👍

arjunrn commented:
:+1:
